### DISCUSSION TITLE
FineTuning pre-trained recognition model on own dataset with vocab change (POLISH)

### DISCUSSION THREAD

**Discussion by boba-fett-hunter** [mindee/doctr#1136] [Q&A]
Hi!

I am trying to train the crnn_vgg16_bn model to recognize the Polish language.

I want to use a pre-trained model for this and show it the Polish language only with a small number of examples.

I use a ready-made script to train:  references/recognition/train_pytorch.py. With the --pretrained (True) parameter.

The model behaves as if it was not fine-tuned, but trained from scratch!. Even if I give him a minimal sample of data, he gets completely lost when evaluating the sample text.
Only giving him 100k samples gives any results.

At first I thought it was related to the size of the modified vocabulary for the Polish language. However, an attempt to use dictionaries for English or French and remove Polish characters gives the same results - model looks like trained from scratch

Please help :)

**Comment by TomekPro** [comment]
Just to add to what my colleasgue described.

I stopped debugger in `doctr/references/recognition/train_pytorch.py:231` when the pretrained model is just loaded:
```model = recognition.__dict__[args.arch](pretrained=args.pretrained, vocab=vocab)```
![image](https://user-images.githubusercontent.com/49069202/215529276-28e99491-4a62-401a-aebf-ea5d76d96fd4.png)

and tried to get a prediction already here - so before any pretraining. Already here the model returns crap - so either loaded model is not pretrained or the picture is somewhat odd - they are padded as in the example below.
```
b = next(iter(val_loader))
out = model(b[0], b[1], return_preds=True)
```
![image](https://user-images.githubusercontent.com/49069202/215529616-307b6746-2d66-44cd-a191-42fed8af4e49.png)

and example image at this point look like this:
![image](https://user-images.githubusercontent.com/49069202/215529776-4aae8c42-f0e6-4eb4-8eee-7655a208f44b.png)

**Comment by TomekPro** [comment]
The problem was connected with the fact that whole head is removed (and randomly initialized) when vocab is different that the one the net was trained on.
![image](https://user-images.githubusercontent.com/49069202/215743650-50e92e39-2343-481a-a254-07ff3a6c95a1.png)

Also I see that when fine-tunning to other language (even for that this is similar to french) whole network is trained - it might be worth to froze backbone part and train only head.

**Reply by krzynio** [reply]
@TomekPro have you fixed it somehow? I just checked the same scenario and receiving a garbage after test synthetic training :)

AC[i0f P0lańd S฿. Z Ó.Ó. 
ńÓ+0Ż6.Kat0W|C€ 
BD0: ÓÓÓÓ6ÓÓ44 
A065 }rU8ZkÓW 
Ś0l|ÓŻf0ŚC| l Ó5-808 P@^Zń{€W 

You can see the pattern though ;)

**Reply by TomekPro** [reply]
yes, vocab for the trainig and at inference time must be the same. So when training you make, for example vocab "polish":
![image](https://user-images.githubusercontent.com/49069202/226325710-57485f64-0216-4a84-a7a0-e69be2f84993.png)

and the at inference as well:
```
        # Load recognition model
        reco_model = crnn_vgg16_bn(pretrained=False, vocab=VOCABS["polish"])
        reco_params = torch.load(recognition_model_path, map_location="cpu")
        reco_model.load_state_dict(reco_params)

        # Instantiate predictor
        self.predictor = ocr_predictor(det_model, reco_model, pretrained=True, preserve_aspect_ratio=False)
```

**Reply by krzynio** [reply]
Hi @TomekPro and thanks for your answer ;)

I have it running without any issues, and the result is almost perfect (besides that  the detection model is merging boxes sometimes) and some lowecase / uppercase issues:

Action Poland sp. Z 0.0. 
40-026.Katowice 
BDO: 000066044 
A065 Pruszków 
Solidarności 1 05-808 Parzniew 

My question was a bit imprecise:

- have you found a way to use models pretrained by mindee and finetune with custom (other) vocabulary without the need of training  whole network  from scratch?

**Reply by TomekPro** [reply]
Hi,
No, we trained recognition model.
